File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func exists (root *TreeNode,p *TreeNode,q *TreeNode, pe *bool,qe *bool)bool{

	if *pe && *qe {return true}
	if root == nil {return *pe && *qe}
	if p==root {*pe = true}
	if q==root {*qe = true}

	left :=exists(root.Left,p,q,pe,qe)
	right:=exists(root.Right,p,q,pe,qe)

	return left ||right


}
func lowestCommonAncestor(root *TreeNode, p *TreeNode, q *TreeNode) *TreeNode {


	if root==nil{return root}
	pe:=false
	qe:=false
	

	left:=lowestCommonAncestor(root.Left,p,q)
	if left != nil {
		return left
	}
	right:=lowestCommonAncestor(root.Right,p,q)
	if right != nil {
		return left
	}
	

	
	if exists(root,p,q,&pe,&qe) {
		return root
	}
	return nil
	
    
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
int count=0;
int helper(TreeNode * root){
    if(!root) return -1;
        int ld = 1 + helper(root->left);
        int rd = 1 + helper(root->right);
        count=max(count,ld+rd);
        return max(rd,ld); 
}
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        helper(root);
        // int ld = 1 + diameterOfBinaryTree(root->left);
        // int rd = 1 + diameterOfBinaryTree(root->right);
        // count=max(count,ld+rd);
        return count; 
    }
};
